File: app/src/main/java/in/kashewdevelopers/airdistance/async_tasks/PlaceToLatLngTask.java
package in.kashewdevelopers.airdistance.async_tasks;

import android.content.Context;
import android.location.Address;
import android.location.Geocoder;
import android.os.AsyncTask;
import android.text.Editable;

import androidx.annotation.NonNull;

import java.lang.ref.WeakReference;
import java.util.List;

import in.kashewdevelopers.airdistance.LocationObject;


@SuppressWarnings("deprecation")
public class PlaceToLatLngTask extends AsyncTask<Editable, Void, LocationObject> {

    public interface OnTaskCompleteListener {
        void onTaskCompleteListener(LocationObject locationObject, String locationType, int taskId);
    }

    OnTaskCompleteListener onTaskCompleteListener;

    private WeakReference<Context> context;
    private String locationType;
    private int taskId;

    public PlaceToLatLngTask(@NonNull Context context, @NonNull String locationType, int taskId) {
        this.context = new WeakReference<>(context);
        this.locationType = locationType;
        this.taskId = taskId;
    }

    @Override
    protected LocationObject doInBackground(Editable... placeList) {
        if (placeList.length <= 0) {
            return null;
        }

        String place = String.valueOf(placeList[0]);

        Geocoder geocoder = new Geocoder(context.get());
        List<Address> addressList;

        try {
            addressList = geocoder.getFromLocationName(place, 10);
        } catch (Exception e) {
            addressList = null;
        }

        if (addressList != null && addressList.size() >= 1) {
            Address address = addressList.get(0);

            if (address.getMaxAddressLineIndex() < 0)
                return null;

            return new LocationObject(address.getAddressLine(0),
                    address.getLatitude(), address.getLongitude());
        }

        return null;
    }

    @Override
    protected void onPostExecute(LocationObject locationObject) {
        super.onPostExecute(locationObject);
        if (onTaskCompleteListener != null) {
            onTaskCompleteListener.onTaskCompleteListener(locationObject, locationType, taskId);
        }
    }

    public PlaceToLatLngTask setOnTaskCompleteListener(@NonNull OnTaskCompleteListener onTaskCompleteListener) {
        this.onTaskCompleteListener = onTaskCompleteListener;
        return this;
    }
}

File: app/src/main/java/in/kashewdevelopers/airdistance/async_tasks/PlaceToLatLngTask.kt
package `in`.kashewdevelopers.airdistance.async_tasks

import `in`.kashewdevelopers.airdistance.data_containers.LocationObject
import android.content.Context
import android.location.Address
import android.location.Geocoder
import android.os.AsyncTask
import android.text.Editable
import java.lang.ref.WeakReference

class PlaceToLatLngTask(context: Context, private val locationType: String, private val taskId: Int) : AsyncTask<Editable, Void, LocationObject>() {

    private val weakReference = WeakReference(context)

    override fun doInBackground(vararg params: Editable?): LocationObject? {
        if (params.isNullOrEmpty())
            return null

        val place: String = params[0].toString()

        val geocoder = Geocoder(weakReference.get())
        val addressList: MutableList<Address>

        try {
            addressList = geocoder.getFromLocationName(place, 10)
        } catch (e: Exception) {
            return null
        }

        if (addressList.isNullOrEmpty()) return null

        val address: Address = addressList[0]
        if (address.maxAddressLineIndex < 0) return null

        return LocationObject(address.getAddressLine(0), address.latitude, address.longitude)
    }

    override fun onPostExecute(result: LocationObject?) {
        super.onPostExecute(result)
        onTaskCompleteListener?.onTaskCompleteListener(result, locationType, taskId)
    }


    interface OnTaskCompleteListener {
        fun onTaskCompleteListener(locationObject: LocationObject?, locationType: String, taskId: Int)
    }

    private var onTaskCompleteListener: OnTaskCompleteListener? = null

    fun setOnTaskCompleteListener(onTaskCompleteListener: OnTaskCompleteListener?): PlaceToLatLngTask {
        this.onTaskCompleteListener = onTaskCompleteListener
        return this
    }

}
